src/Features/Core/Portable/BraceCompletion/AbstractBraceCompletionService.cs

```diff
@@ -0,0 +1,211 @@
+        /// </summary>
+        protected abstract bool IsValidClosingBraceToken(SyntaxToken token);
+
+        public abstract Task<bool> AllowOverTypeAsync(BraceCompletionContext braceCompletionContext, CancellationToken cancellationToken);
+
+        public async Task<BraceCompletionResult?> GetBraceCompletionAsync(BraceCompletionContext braceCompletionContext, CancellationToken cancellationToken)
+        {
+            var closingPoint = braceCompletionContext.ClosingPoint;
+            if (closingPoint < 1)
+            {
+                return null;
+            }
+
+            var openingPoint = braceCompletionContext.OpeningPoint;
+            var document = braceCompletionContext.Document;
+
+            var sourceText = await document.GetTextAsync(cancellationToken).ConfigureAwait(false);
+            if (sourceText[openingPoint] != OpeningBrace)
+            {
+                return null;
+            }
+
+            var root = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            var token = root.FindToken(openingPoint, findInsideTrivia: true);
+            var validOpeningPoint = await IsValidOpenBraceTokenAtPositionAsync(token, openingPoint, document, cancellationToken).ConfigureAwait(false);
+            if (!validOpeningPoint)
+            {
+                return null;
+            }
+
+            var braceTextEdit = new TextChange(TextSpan.FromBounds(closingPoint, closingPoint), ClosingBrace.ToString());
+            // The caret location should be in between the braces.
+            var originalOpeningLinePosition = sourceText.Lines.GetLinePosition(openingPoint);
+            var caretLocation = new LinePosition(originalOpeningLinePosition.Line, originalOpeningLinePosition.Character + 1);
+            return new BraceCompletionResult(ImmutableArray.Create(braceTextEdit), caretLocation);
+        }
+
+        public virtual Task<BraceCompletionResult?> GetTextChangesAfterCompletionAsync(BraceCompletionContext braceCompletionContext, CancellationToken cancellationToken)
+            => SpecializedTasks.Default<BraceCompletionResult?>();
+
+        public virtual Task<BraceCompletionResult?> GetTextChangeAfterReturnAsync(BraceCompletionContext braceCompletionContext, CancellationToken cancellationToken)
+            => SpecializedTasks.Default<BraceCompletionResult?>();
+
+        public virtual async Task<bool> CanProvideBraceCompletionAsync(char brace, int openingPosition, Document document, CancellationToken cancellationToken)
+        {
+            // check that the user is not typing in a string literal or comment
+            var tree = await document.GetRequiredSyntaxTreeAsync(cancellationToken).ConfigureAwait(false);
+            var syntaxFactsService = document.GetRequiredLanguageService<ISyntaxFactsService>();
+
+            return OpeningBrace == brace && !syntaxFactsService.IsInNonUserCode(tree, openingPosition, cancellationToken);
```

Comment:
```suggestion
            return OpeningBrace == brace && !syntaxFactsService.IsInNonUserCode(tree, openingPosition, cancellationToken);
```

do this check first.  before the expensive work to get the tree.  this applies everywhere.  we do this work on every keystroke, so we want to bail out in the common cases as quickly as possibley.